### PR TITLE
TLA+ formalization of Percolator.

### PR DESCRIPTION
**Pull request by foreverbell**
Based on the work of @siddontang https://github.com/siddontang/tlaplus-example/blob/master/Percolator.tla.

This TLA+ spec is verified by TLC with these parameters.
* KEY = 1..5
* CLIENT = {"C1", "C2", "C3"}

Also we've proved TypeInvariant with TLAPS.

### CODE REVIEW

File: Percolator/Percolator.tla [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+  /\ ClientStateTypeInv
+  /\ ClientTsTypeInv
+  /\ ClientKeyTypeInv
+  /\ KeyDataTypeInv
+  /\ KeyLockTypeInv
+  /\ KeyWriteTypeInv
+
+Invariant == TypeInvariant
+
+LEMMA InitStateSatisfiesInvariant ==
```
Comment (siddontang) [inline Percolator/Percolator.tla]: can we move the proof to the end of the file?

Comment (foreverbell) [reply Percolator/Percolator.tla]: Done.

File: Percolator/Percolator.tla [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+\* Cleans up a stale lock when the client encounters one.
+cleanup(c) ==
+  LET
+    start_ts == client_ts[c].start_ts
+    primary == client_key[c].primary
+    secondary == client_key[c].secondary
+  IN
+    \/ /\ hasStaleLock(primary, start_ts)
+       /\ cleanupStaleLock(primary, start_ts)
+    \/ \E k \in secondary :
```
Comment (siddontang) [inline Percolator/Percolator.tla:126]: \A here?

Comment (foreverbell) [reply Percolator/Percolator.tla:126]: Pay attention to the outer "\\/".
This cleans up only one lock, and cleans one by one.. so \E is intended.

File: Percolator/Percolator.tla [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    /\ \A k \in secondary :
+        /\ ~hasLockLE(k, start_ts)
+
+\* Returns TRUE if a lock can be cleanup up.
+\* A lock can be cleaned up iff,
+\*  1. The lock is a primary lock, and its ts is less than or equal to $ts$.
+\*  2. Or the lock is a secondary lock, and the associated primary key (<= ts)
+\*     has been cleaned up.
+isStaleLock(k, l, ts) ==
+  \/ /\ l.primary = k  \* this is a primary lock.
```
Comment (siddontang) [inline Percolator/Percolator.tla]: checking ts at first may be better? like
```
/\ l.ts <= ts
/\ \/ primary
   \/ second 
```

File: Percolator/Percolator.tla [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+  /\ UNCHANGED <<key_lock, key_data, key_write, client_key>>
+
+Get(c) ==
+  /\ client_state[c] = "working"
+  /\ IF canPrewrite(c)
+       THEN
+         /\ client_state' = [client_state EXCEPT ![c] = "prewriting"]
+         /\ UNCHANGED <<next_ts, key_lock, key_data, key_write, client_ts, client_key>>
+       ELSE
+         /\ cleanup(c)
```
Comment (siddontang) [inline Percolator/Percolator.tla:200]: if clean up failed, how can we enter aborted state?

Comment (foreverbell) [reply Percolator/Percolator.tla:200]: I prefer removing the aborted state for simplicity.

If we cannot clean up any locks, the corresponding client will just stay halted, i.e. it has no following states, stutters in the old state.

Comment (siddontang) [reply Percolator/Percolator.tla:200]: This is not a full transaction flow, I suggest moving to the aborted state.

Comment (foreverbell) [reply Percolator/Percolator.tla:200]: As we don't care about liveness property, how about adding one `Abort(c)` action in `Next`? One client can abort at any time if it is not committed, this should complete the transaction workflow, and is what I suggest to do. So I did that.

**Comment by foreverbell** [comment]
I have a follow-up PR to verify the snapshot isolation property. Will send it out once this PR gets merged.

File: Percolator/Percolator.tla (hunk 4 above)
Comment (siddontang) [reply Percolator/Percolator.tla:200]: No need to handle aborted in Next like https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/TCommit.tla does. 

IMO, it is very strange that a transaction can keep the working state but not go to aborted.

Comment (foreverbell) [reply Percolator/Percolator.tla:200]: Discussed offline, agreed on lifting `Abort` as a separate action. Added more comments to describe its meaning.

File: Percolator/Percolator.tla [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+\* $ts$.
+hasLockLE(k, ts) ==
+  \E l \in key_lock[k] : l.ts <= ts
+
+\* Checks whether there is a lock of key $k$ with $ts$.
+hasLockEQ(k, ts) ==
+  \E l \in key_lock[k] : l.ts = ts
+
+\* Returns TRUE if the client encounters no lock, then can do pre-write.
+canPrewrite(c) ==
```
Comment (IANTHEREAL) [inline Percolator/Percolator.tla]: at first glance it make me puzzled, I had to find the definition of `prewrite` to confirm it. 
 it's better to rename other name like `canGoPrewrite` or add some other comments

**Review by IANTHEREAL — COMMENTED** [review]
any other problems？ @siddontang

**Comment by foreverbell** [comment]
Discussed offline, we should also try to clean up stale secondary locks even if the primary lock still exists. This should be more approximate to the implementation.

Added `LockInvariant` to verify there is no missing secondary locks before commit.

**Comment by siddontang** [comment]
should add an abort invariant, if aborted, the primary key can't be committed.

**Comment by foreverbell** [comment]
Done.

**Comment by siddontang** [comment]
Great work. LGTM

**Comment by IANTHEREAL** [comment]
LGTM

**Pull request merged**